make @@toStringTag and constructor prototype properties funky accessors

michaelficarra opened the pull request:
Fixes #286. This includes the `@@toStringTag` change from #213 and also applies the same accessor strategy to "constructor".  This effectively reverts #214.

We're hoping to not have to do this, but if we cannot resolve the web compatibility issue through social mechanisms in a timely manner, this will resolve it through (gross) technical mechanisms.

Jack-Works commented:
oh no... although ses (they're also freezing intrinsics) also use getter/setter to fix web compatibility problems, I never expected it would appear in the spec. I hope we can fix the web compat problem, or at least do the rename like `IterableIterator` instead of this

bakkot reviewed (approved):
spec text LGTM if we have to go this route

bakkot commented:
> at least do the rename like `IterableIterator` instead of this

Renaming would be a _much_ more annoying change for users. No normal programmer will ever notice this change, as inelegant as it appears, whereas everyone will notice a worse name.

ljharb commented on spec.html:
```diff
@@ -289,7 +289,35 @@ <h1>Iterator.prototype</h1>
+          <h1>set Iterator.prototype.constructor</h1>
+          <p>The value of the [[Set]] attribute is a built-in function that takes an argument _v_. It performs the following steps when called:</p>
+          <emu-alg>
+            1. Let _O_ be ? RequireObjectCoercible(*this* value).
+            1. If _O_ is %Iterator.prototype%, then
+              1. Return *undefined*.
+            1. Let _desc_ be ? _O_.[[GetOwnProperty]](*"constructor"*).
+            1. If IsDataDescriptor(_desc_) is *true* and _desc_.[[Writable]] is *false*, then
+              1. Return *undefined*.
+            1. Let _newDesc_ be the PropertyDescriptor { [[Value]]: _v_, [[Writable]]: *true*, [[Enumerable]]: *false*, [[Configurable]]: *true* }.
+            1. Perform ? _O_.[[DefineOwnProperty]](*"constructor"*, _newDesc_).
+            1. Return *undefined*.
```
if we're really going to do this it should be an AO that many things can share, since the only things that potentially change are Iterator.prototype and the property name

michaelficarra replied:
that can be an editorial thing we address in the PR to 262

ljharb reviewed (commented):
oof, the set stuff is very weird

ChristianUlbrich commented:
My take on this, that this is purely a user problem™ . As @ljharb has said:
> Freezing builtins is exceedingly rare, and I'd expect the sites that do to be aggressive in keeping up to date.

It is an _unfortunate_ combination, of combining a _magic library_ offering a _default_ ly enabled _“tamper-free“_ mode with a very outdated library `regenerator-runtime` and current browsers. FWIW, why do they ship it anyway? Generators have been supported since [ages in the browser](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator?retiredLocale=de#browser_compatibility)… (don‘t you dare mentioning something starting with `Int`…)

I do agree, that as the Web is _today_™ , this is not yet shippable, but _if_ we want to go _this_ route (no pun intended), we should do consistently, shouldn‘t we? Meaning, would this not affect other proposals as well? If so, can we make sure, that we can **always** alleviate the **same** problem with _this_ trick?

I have the feeling, that we should try to solve this problem, like we did in the past, by reaching out to the various vendors and web sites, to update their code. The broken library has already been fixed 2 years ago and the airgap.js folks have been contacted. They might implement some blacklist-logic for `runtime-regenerator` (though this might prove difficult, I have to admit).

I know there have been changes in design in the spec in the past to accommodate Web reality, i.e. `[].flatMap()`, but was has been the rationale back then?

This is another use case for a set of TC-39-condoned `eslint` - _rules_, freezing builtins comes at a seemingly high cost and it shows, that there are other valid use cases for ShadowRealms! (cc @caridy @erights @leobalter @legendecas ).

zloirock commented:
How funny it's to read the innocent comment above when only in the last week users wrote to me twice about support Safari 5.1, twice about support IE8, once about IE7, and once about Android 2.3.

ChristianUlbrich commented:
@zloirock Let's do not turn this into a discussion, about on how long a browser should be supported. Companies tend to find an answer to this question, quite fast and sometimes it is just reality, that answers this question.
If somebody wants to use `runtime-regenerator` they can still happily do so, they just should not use a version, that is 2 years old.

bakkot commented:
Regardless of whether people _should_ avoid using old libraries, empirically they do use them, and that is extremely unlikely to change regardless of what we do.

bakkot commented on spec.html:
```diff
@@ -289,7 +289,35 @@ <h1>Iterator.prototype</h1>
+              1. Return *undefined*.
+            1. Let _desc_ be ? _O_.[[GetOwnProperty]](*"constructor"*).
+            1. If IsDataDescriptor(_desc_) is *true* and _desc_.[[Writable]] is *false*, then
+              1. Return *undefined*.
```
Per request of @erights this should be an error (unless the values match), to emulate strict rather than sloppy assignment.

zloirock commented:
> they just should not use a version, that is 2 years old

Oh, I could show you one interesting statistic. Since I mainly work on `core-js`, it's statistics for `core-js` - however, for libraries like `regenerator-runtime` it's even significantly worse. It's downloads for the previous 7 days.

![image](https://github.com/tc39/proposal-iterator-helpers/assets/2213682/437eb23f-9260-4c66-be2e-f196dce00753)

Only 33% of downloaded versions are made in this year. 71% of them are `core-js@3`, but only 33% from this year. Once again, it's statistics for the previous week, not for all time.

Almost all users write something like `"core-js": "^3.0.0"`, but now they have something like `package-lock.json` - and in most cases, users have the version that was actual at the first install time even after many years.

10 years ago I could fix a bug and be sure that ~99% of the users that will download this library the next day will have a version without this bug. Now, after fixing a bug, I can't be sure that at least half of the users who will download this major version of the library next year will not have this bug.

Lock files everywhere - evil -)

michaelficarra commented:
Please bring the off-topic discussion elsewhere. Discussion here should be about this PR.

zloirock commented:
@michaelficarra it's not off-topic - I just show that such alternatives are impossible

![image](https://github.com/tc39/proposal-iterator-helpers/assets/2213682/099b6d13-7f9a-4342-aa87-dbe5f6cbf478)

michaelficarra commented:
@zloirock That's from a discussion thread, not this PR thread. You can add your comments there. Discussion here makes it harder to track what feedback needs to be addressed.

bathos commented:
I’m a bit confused by throwing when the assigned value is %Iterator.prototype%. I assume that’s motivated by the specific web compat issue, but it’s mysterious looking especially given it pivots on the identity of an unbranded object. More broadly, is there a summary somewhere of what the overall behavior is achieving vis a vis the web compat issue?

michaelficarra commented:
@bathos The throw in step 2a is simulating an assignment to a non-writable property in strict mode. Before the recent change to make it throw, it just was just a noop, which is more like an assignment to a non-writable property in sloppy mode. This specific behaviour was requested at the last presentation to TC39.

bathos commented:
@michaelficarra Thanks! I was able to follow it better this time around — though the own-unwritable-"constructor" branch remains opaque to me; I couldn’t figure out for sure why that one’s a special case.

It’s a bit sad that it’s gotta emulate strict-mode assignment behavior specifically, and I think that’s part of what threw me off reading it the first time for whatever reason, but I think I understand the trade offs being made: the alternatives would have been (a) the setter having magical knowing-if-we’re-in-sloppy-mode powers which couldn’t be explained in terms of / implemented using ES code or (b) exoticism, which would actually be explainable in ES code terms and would actually get both sloppy and strict behaviors “right” (so to speak) but which, _well_ ... I imagine nobody wants to `go there` for this lol.

michaelficarra commented:
> having magical knowing-if-we’re-in-sloppy-mode powers which couldn’t be explained in terms of / implemented using ES code

Exactly. So the better choice is to just behave like strict mode in all cases, since we want the environment to be most sane for strict mode code.

bathos commented:
> The setter's behaviour can be described as "assign to the constructor property of the this value as if it had no constructor property in its prototype chain", so we need to first make sure the constructor own-property, if it exists already, is non-writable.

Oh, oops!! I was failing on basic reading comprehension there again and got confused because I’d forgotten there are two setters here and ended up mistaking that unwritable "constructor" step was happening in the @@toStringTag setter! 

> I guess technically we don't trigger an own accessor. Do you think we should?

No, that makes sense the way it is, I think; plenty of stuff calls [[GOP]] and not [[Get]], right? (e.g. HasOwnProperty). However if I’m understanding right (fingers crossed) ... shouldn’t it also care about [[Configurable]]: false? Not necessarily with an explicit step, but more broadly by not ignoring the return value of the [[DefineOwnProperty]] call? Right now it looks like it’ll just return undefined if that returns false (as it [typically] would in the [[Configurable]]: false case).

ljharb commented:
I thought the sentiment of plenary was to simply remove constructor and toStringTag rather than ship them with these weird semantics?

michaelficarra commented:
> shouldn’t it also care about [[Configurable]]: false

Not really, since we shouldn't be trying to change the configurability. But you're right that, at the moment, we may be. Also, we should be throwing when we attempt to create a property on a non-extensible object, and that one would be solved by checking the return value of `[[DefineOwnProperty]]`. Looks like I need to make some updates.

> I thought the sentiment of plenary was to simply remove constructor and toStringTag rather than ship them with these weird semantics?

That was briefly brought up as an option, yes. We haven't decided what approach we will take if we end up needing to take action (we still hope we won't have to), but we want this one to be ready to go in case we opt for it.

michaelficarra commented:
@bathos @bakkot I've rewritten the setter. It now matches assignment better when pulled off and `.call`-ed, and re-uses existing AOs more. I think it also fixed a bug where the property was being created non-enumerable before. Please review for correctness.

bathos commented on spec.html:
```diff
@@ -107,6 +107,30 @@ <h1>Well-Known Intrinsic Objects</h1>
+  <emu-clause id="sec-SetterThatIgnoresPrototypeProperties" type="abstract operation">
+    <h1>
+      SetterThatIgnoresPrototypeProperties (
+        _home_: an Object,
```
Is it intentional for this parameter name to evoke [[HomeObject]]? The phrasing “On the home_object ” in particular strongly suggests that to me.

It does kind of of align with an intuitive notion of what the hypothetical [[HomeObject]] values of these setters could have been set to (i.e., if built-in functions had that slot). The connection is pretty tenuous though — ES code polyfills for these functions, for example, would be very unlikely to end up with %Iterator.prototype% as their “real” [[HomeObject]] values. Being a merely near conceptual match actually seems to me like a good reason not to use the name. (Maybe something like “intrinsicOwner” or even just “proto”?)

The naming of an AO variable isn’t very important, of course — just a thought I had when reading the updated text (which seems much clearer now btw).

michaelficarra replied:
I was aware of the non-uniqueness of the name, but when I asked myself "what is the name of the place where this property lives?", the obvious answer was "home". That is probably also how `[[HomeObject]]` got its name. But I don't think the alias name alone will imply a strong relationship to `[[HomeObject]]`. I'm open to alternative suggestions if you've got them.

bathos commented on spec.html:
```diff
@@ -107,6 +107,30 @@ <h1>Well-Known Intrinsic Objects</h1>
+        1. Throw a *TypeError* exception.
+      1. Let _desc_ be ? _O_.[[GetOwnProperty]](_p_).
+      1. If _desc_ is *undefined*, then
+        1. Perform ? CreateDataPropertyOrThrow(_O_, _p_, _v_).
```
I figured the unenumerability in the previous iteration was intentional given typical "constructor" and @@toStringTag properties is enumerable and the accessor property itself won’t be either, but I agree this is likely the better behavior due to to being less unique / more similar to OrdinarySet.

michaelficarra replied:
Yeah it looks like the override mistake does not cause the new property to inherit the enumerability or configurability of the existing property on the prototype. So making it enumerable/writable/configurable is definitely the correct thing to do.

bakkot commented on spec.html:
```diff
@@ -107,6 +107,30 @@ <h1>Well-Known Intrinsic Objects</h1>
+      1. If _desc_ is *undefined*, then
+        1. Perform ? CreateDataPropertyOrThrow(_O_, _p_, _v_).
+      1. Else,
+        1. Perform ? Set(_O_, _p_, _v_, *true*).
```
This will perform `_O_.[[GetOwnProperty]]` a second time, which is pretty weird - are you sure that's what you want? I'm pretty sure that's observable with a Proxy.

What if we just throw in this case instead? You have to either be pulling this off or using `Reflect.set` to hit this branch (right?), and I feel like we should just not support those cases instead of trying to figure out the "right" behavior.

michaelficarra replied:
The second `[[GetOwnProperty]]` is not ideal, and could theoretically be avoided by extending the MOP, but I don't think we should do that, and I'm not too worried about it.

bakkot reviewed (approved):
This basically matches [what SES/node does](https://github.com/nodejs/node/blob/ed16a46481bd3d4fd05ec568d2bffeff492c58ad/lib/internal/freeze_intrinsics.js#L499-L517), except in their case they're behaving as if frozen and therefore throw in the "this === object" case. So, sure.

anba commented on spec.html:
```diff
@@ -107,6 +107,30 @@ <h1>Well-Known Intrinsic Objects</h1>
+    <dl class="header">
+    </dl>
+    <emu-alg>
+      1. Let _O_ be ? RequireObjectCoercible(*this* value).
```
The `this value` needs to be passed as an argument to an abstract operation, because the term `this value` only applies to functions defined after [ch. 18](https://tc39.es/ecma262/#sec-ecmascript-standard-built-in-objects). 

(The definitions for `this value` and `NewTarget` in ch. 18 seem to be kind of broken, because they refer to [ch. 10.3](https://tc39.es/ecma262/#sec-built-in-function-objects), which no longer defines these terms. For example compare to [ES 2020](https://262.ecma-international.org/11.0/#sec-built-in-function-objects), where `this value` and `NewTarget` are still mentioned in the "Built-in Function Objects" text.)

anba commented on spec.html:
```diff
@@ -107,6 +107,30 @@ <h1>Well-Known Intrinsic Objects</h1>
+    <dl class="header">
+    </dl>
+    <emu-alg>
+      1. Let _O_ be ? RequireObjectCoercible(*this* value).
```
The next lines seems to be assume `O` is an object, so I guess this should actually be `ToObject(thisValue)` instead of `RequireObjectCoercible`.

michaelficarra replied:
https://github.com/tc39/proposal-iterator-helpers/pull/290
